File: binary_search_tree.py
class CreateNode:
    def __init__(self, arr, mid):
        self.data = arr[mid]
        self.left = self.right = None


def create_bst(arr, left, right):
    if left <= right:
        mid = int((left + right) / 2)
        # print(mid)
        root = CreateNode(arr, mid)
        root.left = create_bst(arr, left, mid - 1)
        root.right = create_bst(arr, mid + 1, right)
        return root
    else:
        return None


def inorder(root):
    if root:
        inorder(root.left)
        print(root.data, end='')
        inorder(root.right)


array = [1, 2, 3, 4, 5, 6, 7, 8, 9]
node = create_bst(array, 0, len(array) - 1)
inorder(node)

File: Programs/Trees/BinarySearchTree.java
class Node { 
	
	int data; 
	Node left, right; 
	
	Node(int d) { 
		data = d; 
		left = right = null; 
	} 
} 

class BinarySearchTree { 
	
	static Node root; 

	Node sortedArrayToBST(int arr[], int start, int end) { 

		if (start > end) { 
			return null; 
		} 

		int mid = (start + end) / 2; 
		Node node = new Node(arr[mid]); 
		node.left = sortedArrayToBST(arr, start, mid - 1); 
		node.right = sortedArrayToBST(arr, mid + 1, end); 		
		return node; 
	} 

	void preOrder(Node node) { 
		if (node == null) { 
			return; 
		} 
		System.out.print(node.data + " "); 
		preOrder(node.left); 
		preOrder(node.right); 
	} 
	
	public static void main(String[] args) { 
		BinarySearchTreeTree tree = new BinarySearchTree(); 
		int arr[] = new int[]{1, 2, 3, 4, 5, 6, 7}; 
		int n = arr.length; 
		root = tree.sortedArrayToBST(arr, 0, n - 1); 
		System.out.println("Preorder traversal"); 
		tree.preOrder(root); 
	} 
} 
 
